src/ConnectedMode/Suppressions/SuppressionIssueStoreUpdater.cs

```diff
@@ -30,16 +37,121 @@ internal interface ISuppressionIssueStoreUpdater
         /// <summary>
         /// Fetches all available suppressions from the server and updates the server issues store
         /// </summary>
-        void FetchAllServerSuppressions();
+        Task UpdateAllServerSuppressionsAsync();
 
         /// <summary>
         /// Fetches suppressions from the server from the specified timestamp onwards and updates the issues store
         /// </summary>
-        void FetchServerSuppressions(DateTimeOffset fromTimestamp);
+        Task UpdateServerSuppressionsAsync(DateTimeOffset fromTimestamp);
 
         /// <summary>
         /// Clears all issues from the store
         /// </summary>
         void Clear();
     }
+
+    [Export(typeof(ISuppressionIssueStoreUpdater))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    internal class SuppressionIssueStoreUpdater : ISuppressionIssueStoreUpdater, IDisposable
+    {
+        private readonly ISonarQubeService server;
+        private readonly IServerQueryInfoProvider serverQueryInfoProvider;
+        private readonly IServerIssueStoreWriter storeWriter;
+        private readonly ILogger logger;
+        private readonly IThreadHandling threadHandling;
+
+        private CancellationTokenSource updateAllCancellationTokenSource = new CancellationTokenSource();
+
+        [ImportingConstructor]
+        public SuppressionIssueStoreUpdater(ISonarQubeService server,
+            IServerQueryInfoProvider serverQueryInfoProvider,
+            IServerIssueStoreWriter storeWriter,
+            ILogger logger)
+            : this(server, serverQueryInfoProvider, storeWriter, logger, ThreadHandling.Instance)
+        {
+        }
+
+        internal /* for testing */ SuppressionIssueStoreUpdater(ISonarQubeService server,
+            IServerQueryInfoProvider serverQueryInfoProvider,
+            IServerIssueStoreWriter storeWriter,
+            ILogger logger,
+            IThreadHandling threadHandling)
+        {
+            this.server = server;
+            this.serverQueryInfoProvider = serverQueryInfoProvider;
+
+            this.storeWriter = storeWriter;
+            this.logger = logger;
+            this.threadHandling = threadHandling;
+        }
+
+        #region ISuppressionIssueStoreUpdater
+
+        public async Task UpdateAllServerSuppressionsAsync()
+        {
+            await threadHandling.SwitchToBackgroundThread();
+
+            try
+            {
+                logger.WriteLine(Resources.Suppressions_Fetch_AllIssues);
+
+                CancelCurrentOperation();
```

Comment:
The old class had this logic - we only want to have one "fetch all" operation running, so any time we start one, cancel the previous one if there is one.

However, I don't think the logic was implemented correctly before, and I couldn't find any tests for it. This version is at least tested.